torchdata/dataloader2/utils/worker.py

```diff
@@ -73,6 +103,45 @@ def process_init_fn(
     return datapipe
 
 
+def _set_global_random_state(seed_generator: torch.Generator) -> None:
+    py_seed = generate_random_int(seed_generator)
+    random.seed(py_seed)
+
+    torch_seed = generate_random_int(seed_generator)
+    torch.manual_seed(torch_seed)
+
+    if HAS_NUMPY:
+        # Numpy only accepts uint32 as the seed
+        np_seed = generate_random_int(seed_generator, torch.int32)
+        if np_seed < 0:
+            np_seed = 2 ** 32 + np_seed
+        numpy.random.seed(np_seed)
+
+
+def dispatch_process_reset_fn(
+    datapipe: DataPipe,
+    worker_info: WorkerInfo,
+    dist_info: _DistInfo,
+) -> DataPipe:
+    r"""
+    Based on the distributed shared random seed, this function is used to set the random state
+    of the non-repliable ``DataPipe`` graph and the global random states for the dispatch process.
+    This function would guarantee that all distributed non-sharding process share the
+    same random states to ensure the same shuffle order.
+    """
+    worker_seed_generator = torch.Generator()
+    worker_seed_generator.manual_seed(dist_info.shared_seed)
+    torch.utils.data.graph_settings.apply_random_seed(
+        datapipe,
+        worker_seed_generator,
+    )
+
+    # Set global random states
+    _set_global_random_state(worker_seed_generator)
+
+    return datapipe
```

Comment:
Review Step 9: When new epoch starts, we want to control the random seed based on distributed information.
We need to guarantee all distributed dispatching process share the same random seed.